modules/heat_conduction/src/userobjects/SelfShadowSideUserObject.C

```diff
@@ -0,0 +1,410 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "SelfShadowSideUserObject.h"
+#include "RotationMatrix.h"
+
+#include "libmesh/parallel_algebra.h"
+
+registerMooseObject("HeatConductionApp", SelfShadowSideUserObject);
+
+InputParameters
+SelfShadowSideUserObject::validParams()
+{
+  InputParameters params = SideUserObject::validParams();
+  params.addClassDescription("Compute the illumination status for a self shadowing sideset");
+  params.addRequiredRangeCheckedParam<std::vector<PostprocessorName>>(
+      "illumination_flux",
+      "illumination_flux_size>0 && illumination_flux_size<=3",
+      "Radiation direction vector. Each component of the vector can be a constant number or a "
+      "postprocessor name (the latter enables time varying radiation directions - spatial "
+      "variation is not supported)");
+  return params;
+}
+
+SelfShadowSideUserObject::SelfShadowSideUserObject(const InputParameters & parameters)
+  : SideUserObject(parameters),
+    _dim(_mesh.dimension()),
+    _raw_direction(coupledPostprocessors("illumination_flux")),
+    _illumination_status(
+        declareRestartableData<std::map<SideIDType, unsigned int>>("illumination_status"))
+{
+  // we should check the coordinate system (i.e. permit only 0,0,+-1 for RZ)
+
+  // check problem dimension
+  if (_dim != 2 && _dim != 3)
+    mooseError("SelfShadowSideUserObject works only for 2 and 3 dimensional problems.");
+
+  // fetch raw direction
+  for (const auto i : index_range(_raw_direction))
+    _raw_direction[i] = &getPostprocessorValue("illumination_flux", i);
+}
+
+void
+SelfShadowSideUserObject::initialize()
+{
+  // delete list of collected lines or triangles
+  _lines.clear();
+  _triangles.clear();
+
+  // delete local qps
+  _local_qps.clear();
+
+  // update direction and rotation matrix
+  RealVectorValue direction;
+  for (const auto i : index_range(_raw_direction))
+    direction(i) = *_raw_direction[i];
+  _rotation =
+      _dim == 2 ? RotationMatrix::rotVec2DToX(direction) : RotationMatrix::rotVecToZ(direction);
+}
+
+void
+SelfShadowSideUserObject::execute()
+{
+  const SideIDType id(_current_elem->id(), _current_side);
+
+  // add triangulated sides to list
+  if (_dim == 2)
+    addLines(id);
+  else
+    addTriangles(id);
+
+  // save off rotated QP coordinates got local sides
+  _local_qps.emplace_back(id, _q_point);
```

Comment:
I don't understand how consistency in the usage of displaced/undisplaced mesh coordinates takes place for normals (BC) and shadowing (UO)... How does that work?

Reply:
If we want to enforce consistency I could add an API to the user object to return the use displaced mesh bool. Good idea.